controllers/topology/internal/mergepatch/managed_paths.go

```diff
@@ -0,0 +1,138 @@
+	if err := storeManagedPaths(objWithManagedFieldAnnotation); err != nil {
+		return nil, err
+	}
+	return objWithManagedFieldAnnotation, nil
+}
+
+// getManagedPaths infers the list of paths managed by the topology controller in the previous patch operation
+// by parsing the value of the managed field annotation.
+// NOTE: if for any reason the annotation is missing, the patch helper will fall back on standard
+// two-way merge behavior.
+func getManagedPaths(obj client.Object) []contract.Path {
+	// Gets the managed field annotation from the object.
+	managedFieldAnnotation := obj.GetAnnotations()[clusterv1.ClusterTopologyManagedFieldsAnnotation]
+
+	// Parses the managed field annotation value into a list of paths.
+	// NOTE: we are prepending "spec" to the paths to restore the correct absolute path inside the object.
+	managedFields := strings.Split(managedFieldAnnotation, managedFieldSeparator)
+	paths := make([]contract.Path, 0, len(managedFields))
+	for _, managedField := range managedFields {
+		if strings.TrimSpace(managedField) == "" {
+			continue
+		}
+
+		path := contract.Path{"spec"}
+		for _, f := range strings.Split(managedField, managedPathSeparator) {
+			path = append(path, strings.TrimSpace(f))
+		}
+		paths = append(paths, path)
+	}
+	return paths
+}
+
+// Stores the list of paths managed by the topology controller into the managed field annotation.
+// NOTE: The topology controller is only concerned about managed paths in the spec; given that
+// we are dropping spec. from the result to reduce verbosity of the generated annotation.
+// NOTE: Managed paths are relevant only for unstructured objects where it is not possible
+// to easily discover which fields have been set by templates + patches/variables at a given reconcile;
+// instead, it is not necessary to store managed paths for typed objets (e.g. Cluster, MachineDeployments)
+// given that the topology controller explicitly sets a well-known, immutable list of fields at every reconcile.
+func storeManagedPaths(obj client.Object) error {
+	// Return early if the object is not unstructured.
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return nil
+	}
+
+	// Gets the object spec.
+	spec, ok, err := unstructured.NestedMap(u.UnstructuredContent(), "spec")
+	if err != nil {
+		return errors.Wrap(err, "failed to get object spec")
+	}
+
+	// Build a string representation of the paths under spec which are being managed by the object (the fields
+	// a topology is expressing an opinion/value on).
+	managedFields := ""
+	if ok {
+		s := []string{}
+		paths := paths([]string{}, spec)
+		for _, p := range paths {
+			s = append(s, strings.Join(p, managedPathSeparator))
+		}
+
+		// Sort paths to get a predictable order (useful for readability and testing, not relevant at parse time).
+		sort.Strings(s)
+
+		// Concatenate all the paths in a single string.
+		// NOTE: add an extra space between fields for better readability (not relevant at parse time).
```

Comment:
I don't have strong opinions on the whitespaces, it just makes things more readable.
I'm also open to considering alternatives to the comma-separated list, do you have something in mind?

Reply:
One idea to optimize the footprint slightly would be to aggregate leaf nodes like this (with `{}`: (new lines are just here for illustration and shouldn't be part of the annotation)
```yaml
kubeadmConfigSpec.clusterConfiguration.apiServer.certSANs,
kubeadmConfigSpec.clusterConfiguration.imageRepository,
kubeadmConfigSpec.initConfiguration.nodeRegistration.criSocket,
kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs.{eviction-hard,cgroup-driver}
kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
machineTemplate.infrastructureRef.{apiVersion,kind,name,namespace}
machineTemplate.metadata.labels.{cluster.x-k8s.io/cluster-name,topology.cluster.x-k8s.io/owned},
replicas,
rolloutStrategy.rollingUpdate.maxSurge,
rolloutStrategy.type,
version
```

Of course this would also work on all layers:
```yaml
kubeadmConfigSpec.{clusterConfiguration.{apiServer.{certSANs},imageRepository,},initConfiguration{nodeRegistration.{criSocket,kubeletExtraArgs.{eviction-hard,cgroup-driver}}}
...
```

While the first option os still readable the second one gets quite obscure and it's probably not worth the risk  to introduce bugs.

Reply:
P.S. I couldn't figure out a case where it fails, but can you think of a potential bug because dots in labels cannot be differentiated from the dots of nested fields?

Reply:
we are including only field names in the label value, so there is no concern about undesired dots being added

Reply:
No my point is more like, we are (currently) adding names of individual labels and those labels can contain dots. I was just wondering if that's an issue because we can't know if it's just a dot in a label or one of the nested field

E.g.:
```yaml
 kubeadmConfigSpec.clusterConfiguration.apiServer.certSANs,
        kubeadmConfigSpec.clusterConfiguration.imageRepository, kubeadmConfigSpec.initConfiguration.nodeRegistration.criSocket,
        kubeadmConfigSpec.initConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.cgroup-driver,
        kubeadmConfigSpec.joinConfiguration.nodeRegistration.kubeletExtraArgs.eviction-hard,
        machineTemplate.infrastructureRef.apiVersion, machineTemplate.infrastructureRef.kind,
        machineTemplate.infrastructureRef.name, machineTemplate.infrastructureRef.namespace,
        machineTemplate.metadata.labels.cluster.x-k8s.io/cluster-name, machineTemplate.metadata.labels.topology.cluster.x-k8s.io/owned,
        replicas, rolloutStrategy.rollingUpdate.maxSurge, rolloutStrategy.type, version
```

> machineTemplate.metadata.labels.topology.cluster.x-k8s.io/owned,

> machineTemplate.metadata.labels.
Those dots are all from the field

> topology.cluster.x-k8s.io/owned,
Those dots are just part of the label name.

I would have thought that this could lead to issues when keeping/dropping labels, but it doesn't seem to.

Reply:
confirmed, this is not a problem.

Reply:
I played around with it some more and was able to find the issue.

It turns it only did work because I dropped a label with dots from a path which was an authoritativePath anyway (spec.machineTemplate.metadata.labels). So it did only work because it was below spec.machineTemplate.metadata and not because of managed fields.

If I try to do this with some other path it fails:
https://github.com/kubernetes-sigs/cluster-api/pull/5846/commits/f6eee6f872d2177d1c5bba734cb0a0be24f87b11

Reply:
I think we can only fix this by:
* using characters which are not valid character in labels as separator (dot is valid in labels, not sure about commas) or
* escaping labels (dots, maybe also commas)